src/lem/mod.rs

```diff
@@ -1,361 +1,267 @@
-    pub fn new(input: [&str; 3], lem_op: LEMOP) -> Result<LEM> {
-        lem_op.check()?;
-        let dmap = DashMap::from_iter(input.map(|i| (i.to_string(), i.to_string())));
+    /// Performs the static checks described in `LEM`'s docstring.
+    pub fn check(&self) {
+        // TODO
+    }
+
+    /// Instantiates a `LEM` with the appropriate transformations to make sure
+    /// that constraining will be smooth.
+    pub fn new(input: [AString; 3], lem: &LEMCTL) -> Result<LEM> {
+        let mut map = HashMap::from_iter(
+            input
+                .iter()
+                .map(|i| (MetaPtr(i.clone()), MetaPtr(i.clone()))),
+        );
         Ok(LEM {
-            input: input.map(|i| i.to_string()),
-            lem_op: lem_op.deconflict("", &dmap)?,
+            input_vars: input,
+            ctl: lem.deconflict(&Path::default(), &mut map)?,
         })
     }
+
+    /// Intern all symbols that are matched on `MatchSymbol`s
+    #[inline]
+    pub fn intern_matched_symbols<F: LurkField>(&self, store: &mut Store<F>) {
+        self.ctl.intern_matched_symbols(store);
+    }
+
+    /// Asserts that all paths were visited by a set of frames. This is mostly
+    /// for testing purposes.
+    pub fn assert_all_paths_taken<F: LurkField>(
+        &self,
+        frames: &Vec<Frame<F>>,
+        store: &mut Store<F>,
+    ) {
+        assert_eq!(
+            self.ctl.num_paths_taken(frames, store).unwrap(),
+            self.ctl.num_paths()
+        );
+    }
 }
 
 #[cfg(test)]
 mod tests {
+    use super::circuit::SlotsCounter;
     use super::{store::Store, *};
     use crate::{lem, lem::pointers::Ptr};
-    use bellperson::util_cs::test_cs::TestConstraintSystem;
+    use bellperson::util_cs::{test_cs::TestConstraintSystem, Comparable, Delta};
     use blstrs::Scalar as Fr;
 
-    fn constrain_test_helper(lem: &LEM, store: &mut Store<Fr>, witnesses: &Vec<Witness<Fr>>) {
-        for w in witnesses {
+    fn constrain_test_helper(lem: &LEM, exprs: &[Ptr<Fr>], expected_num_slots: SlotsCounter) {
+        let slots_count = lem.ctl.count_slots();
+
+        assert_eq!(slots_count, expected_num_slots);
+
+        let computed_num_constraints = lem.num_constraints(&slots_count);
+
+        let mut store = Store::default();
+
+        let mut cs_prev = None;
+        for expr in exprs {
+            let frames = lem.eval(*expr, &mut store).unwrap();
+
             let mut cs = TestConstraintSystem::<Fr>::new();
-            lem.constrain(&mut cs, store, w).unwrap();
+            for frame in frames.clone() {
+                lem.synthesize(&mut cs, &mut store, &slots_count, &frame)
+                    .unwrap();
+            }
+
             assert!(cs.is_satisfied());
+
+            assert_eq!(computed_num_constraints, cs.num_constraints());
+
+            if let Some(cs_prev) = cs_prev {
+                assert_eq!(cs.delta(&cs_prev, true), Delta::Equal);
+            }
```

Comment:
As I read it, this check and/or the test generally seem to hide some assumptions. I think that in the general case we should not expect this assertion to be true. If this test passes, my guess is that it is only incidentally because the expressions passed to the helper *incidentally* require the same number of iterations.

This is not only a quibble, though: it seems to undermine the purpose of the test. If one point of the test is (as it seems) to ensure circuit uniformity, then this incidental similarity required for expressions to pass the test is also implicitly filtering out cases in which non-uniformity might be expected to manifest.

In other words, evaluations which require exactly the same number of iterations are much less likely to exercise sufficiently different paths during the total evaluation to uncover any latent non-uniformity. This may not be an issue with the current, very restricted, set of test cases, but it is worth noting for when a richer set of possibilities exists.

I believe it's the case that `frames.len()` is always one — and that this is why the issue can't easily be highlighted. If that's true, it might be worth adding an assertion to that effect along with a comment that this won't be true in general. That would document the assumptions/limitations and help ensure that future test implementers are cued to produce a broader variety of evaluation paths within each test case.

Reply:
The evaluation of each expression separately indeed only take one iteration. But note that we are feeding the test function with a list of inputs that are supposed to be evaluated independently. And the union of all resulting frames shouldn't affect uniformity. For this, we have to provide input expressions that are handicrafted to follow different paths inside the LEM

Reply:
What do you mean by the 'union of frames'? I think you mean the concatenation of frames. If you synthesize multiple frames sequentially, the circuit will grow.

As I said, it may be the case that your current tests (and LEM) assume that there will only be one frame — but this is not a good general assumption. I think maybe your circuit will fail due to duplicated namespace paths when you do try to synthesize a circuit for more than one frame.

In any case, if the 'union of frames' means that somehow multiple frames are collapsed into one — then yes, you would have uniformity. However, I think you would then be compressing the proof into something trival — improving prover performance to constant time but at the cost of soundness.

Reply:
Let's use this test case as an example:
```rust
#[test]
fn test_hash_slots() {
    let lem = lem!(expr_in env_in cont_in {
        let x: Cons = hash2(expr_in, env_in);
        let y: Cons = hash3(expr_in, env_in, cont_in);
        let z: Cons = hash4(expr_in, env_in, cont_in, cont_in);
        let t: Terminal;
        let p: Nil;
        match_tag expr_in {
            Num => {
                let m: Cons = hash2(env_in, expr_in);
                let n: Cons = hash3(cont_in, env_in, expr_in);
                let k: Cons = hash4(expr_in, cont_in, env_in, expr_in);
                return (m, n, t);
            },
            Char => {
                return (p, p, t);
            },
            Cons => {
                return (p, p, t);
            },
            Nil => {
                return (p, p, t);
            }
        };
    })
    .unwrap();

    constrain_test_helper(
        &lem,
        &[Ptr::num(Fr::from_u64(42)), Ptr::char('c')],
        SlotsCounter::new((2, 2, 2)),
    );
}
```
We feed the test function with two (unrelated) inputs: the number `42` and the character `'c'`. The evaluation of each of them will produce a single frame, each. The assertion that we care about is that those frames, will generate uniform circuits. That's what `constrain_test_helper` asserts

Reply:
Yes, but the helper is written in such a way that it will fail if one of the expressions produces more than a single frame — while being structured such that it *appears* that the intent is to allow for multi-frame evaluations (because you iterate over the frames and synthesize each in the same constraint system).

I believe it's right for the helper to do that, but if it does then you will *not* end up with uniform circuits except in the case that the expressions provided happen to require the same number of frames.

You may argue that it is the test-writer's obligation to ensure that expressions passed to the helper always *will* result in the same number frames when evaluated. However, as I said originally, this then will exclude the most interesting sets of uniformity — which are those that test expressions following sufficiently different evaluation paths that non-uniformity might be expected to appear (under reasonable expectations of how the code might go wrong — and which these tests would be designed to rule out).

Reply:
Nice catch. I didn't intend to support multi-frame evaluations yet. We can change `synthesize` to always produce a new constraint system from scratch and leave that for later. Is that okay?

Reply:
Synthesize shouldn't create constraint systems. I think you can change your test to not pass the same constraint system to `synthesize` multiple times, if that's your intention. Since I *believe* you never produce multiple frames in these tests, I think it would be easier to assert that than to bend over backward to show that each of the hypothetical multiple frames is uniform (for now). But I think that's something you *should* do later — since it's a stronger assertion than that the composite of some same-numbered-of-frames circuits are the same. The latter tests inter-multiframe uniformity, while the former tests intra-multiframe uniformity. Both are useful and important in terms of what I believe you are and should be trying to demonstrate.

Reply:
Oh I think this was a simple mistake of not creating a new CS for each frame, and instead appending to the circuit (which would create an enormous circuit). Thanks for solving this @emmorais